src/dmangle.d

```diff
@@ -163,14 +172,95 @@ private void MODtoDecoBuffer(OutBuffer* buf, MOD mod)
 extern (C++) final class Mangler : Visitor
 {
     alias visit = super.visit;
+
 public:
+    version(useBackref)
+    {
+        // cannot put C++ classes into a hash table, tries to call toHash
+        size_t[void*] types;
+        size_t[void*] symbols;
+        size_t[void*] idents;
+    }
     OutBuffer* buf;
 
     extern (D) this(OutBuffer* buf)
     {
         this.buf = buf;
     }
 
+    final bool backrefType(Type t)
+    {
+        version(useBackref)
+        {
+            if (!t.isTypeBasic())
+            {
+                if (auto p = cast(void*)t in types)
+                {
+                    buf.printf("#%d", cast(int)*p);
+                    return true;
+                }
+                types[cast(void*)t] = types.length;
+            }
+        }
+        return false;
+    }
+
+    final bool backrefSymbol(Dsymbol s)
+    {
+        version(useBackref)
+        {
+            if (auto p = cast(void*)s in symbols)
+            {
+                buf.printf("$%d", cast(int)*p);
+                return true;
+            }
+            symbols[cast(void*)s] = symbols.length;
```

Comment:
I thought about that, too. Could work, but again, we'll have to revisit the grammar for ambiguities. The type/symbol index yields smaller numbers and might be a bit more readable.